### ISSUE TITLE
network implementation differences (InstanceNorm, probability mean, max-pooling)

### ISSUE THREAD

**Issue by tarepan** [hujinsen/StarGAN-Voice-Conversion#4]
Thanks for your great implementation.  
I found some implementation differences compared to original article.  

# difference list
* IN as alternatives of BN
* mean as alternatives of product in D/C last layer
* (in only C) max-pooling as alternatives of strided-Conv

## IN as alternatives of BN
In Generator, Discriminator and Classifier, Instance Normalization (IN) are used as alternatives of Batch Normalization (BN) ([code](https://github.com/hujinsen/StarGAN-Voice-Conversion/blob/1cb924112b4a594a4605d117905ccfa47c981373/module.py#L11)).  
There are comment-outed Batch Normalization, so are there any problem in BN?  

## mean as alternatives of product in D/C last layer
In original article, probabilities (probability of each patches) is multiplied (== product).  

> the final output D(y, c) is given by the product of all these probabilities.  
> ...
> “Product” denote ... product pooling layers,

But in this implementation, in last layer, probabilities is taken average.  

> c1_red = tf.reduce_mean(c1, keepdims=True) [code](https://github.com/hujinsen/StarGAN-Voice-Conversion/blob/00abc81acddda348188574355f9d4e5b0284b32c/module.py#L239)

Is this intended implementation based on your experiments, or some other reasons?

## (in only C) max-pooling as alternatives of strided-Conv
In D, down-sampling is made by strided-Conv as same as original article.  
But in C, down-sampling is made by max-pooling.  
Why are they used in different manner?

**Comment by hujinsen** [comment]
Thanks for your comment. 

- IN as alternatives of BN

BN normalizes the features by the mean and variance computed within a (mini-)batch. The elements in each channel are normalized by the same mean and variance, this may cause the different speaker ids that attached to the layers  mixed , thus we loose speaker ids information. By the way, if the original paper  set batch size to 1, because there is just one speaker id information attached(can’t mix), the net will also works. In the  implementation, batch size is 8,in order to avoid the problem , so I  choose instance norm instead. I haven’t tried the layer norm or group norm.

- mean as alternatives of product in D/C last layer

I’m sorry aboult this problem because I don’t konw what the “product pooling layer” is. This problem confused me a lot, if you know how to implement this layer please notify me.

- (in only C) max-pooling as alternatives of strided-Conv

Sorry about this too.The paper uses strided-Conv and have no padding information of each layer. Due to CNN’s trans constant, I found the classifier worked just ok.

Hopefully to  answer your issue completely. Any advice are welcome!

**Comment by tarepan** [comment]
Thanks for your kind answers.  

## IN as alternatives of BN
I have not thought about BN problem. I understand, thanks!!  

## mean as alternatives of product in D/C last layer
I interpret "product pooling layer" as simply multiplying all elements.  
In StarGAN-VC D,  
(batch, 1-channel,  1-height,  64-width(w0 ~ w63))  
=>  
(batch, 1-channel,  1-height,  1-width(w0 * w1 * w2 * ... * w63))  
In my case (PyTorch), simply implement as  
```
# in forward
out = input[:, :, :, 0]
for i in range(1, input.size()[3]) # width length
  out = out *  input[:, :, :, i]
return out
```
But this code is ugly and I do not know whether back prop is properly go through this process...

## (in only C) max-pooling as alternatives of strided-Conv
I agree with you, some padding information is missing but some unusual padding seems to be used (dimension is curious in C).  
Sorry, I cannot catch up some part of what you said.  

> Due to CNN’s trans constant,  

what it mean?

**Comment by hujinsen** [comment]
Sorry about the confusion  words. I mean CNN is "translation invariance" , so if doesn't matter that we choose other  padding size or filter size as long as they are reasonable. 

Your implementation of "product pooling layer" , multiplying all the elements, if one elements is zero then the result is zero. maybe the network is not intended to operate like this.  I don't know if I'm right, but the problem should be taken into consideration.
